### PR TITLE
FS: Allow multiple instances of the same archive type to be open at once

### PR DESCRIPTION
**Pull request by yuriks**
There's some code related to gamecoin that I had to disable since it wouldn't work with the new stuff. @Subv said he's going to fix it.

### CODE REVIEW

File: src/core/file_sys/archive_backend.h [hunk 1]
```diff
@@ -260,4 +246,29 @@ class ArchiveBackend : NonCopyable {
     virtual std::unique_ptr<DirectoryBackend> OpenDirectory(const Path& path) const = 0;
 };
 
+class ArchiveFactory : NonCopyable {
+public:
+    virtual ~ArchiveFactory() {
+    }
+
+    /**
+    * Get a descriptive name for the archive (e.g. "RomFS", "SaveData", etc.)
```
Comment (Subv) [inline src/core/file_sys/archive_backend.h]: Add a space before the * for every comment line :)

File: src/core/file_sys/archive_romfs.cpp [hunk 2]
```diff
@@ -15,11 +15,24 @@
 
 namespace FileSys {
 
-Archive_RomFS::Archive_RomFS(const Loader::AppLoader& app_loader) {
+ArchiveFactory_RomFS::ArchiveFactory_RomFS(const Loader::AppLoader& app_loader)
+        : romfs_data(std::make_shared<std::vector<u8>>()) {
     // Load the RomFS from the app
-    if (Loader::ResultStatus::Success != app_loader.ReadRomFS(raw_data)) {
+    if (Loader::ResultStatus::Success != app_loader.ReadRomFS(*romfs_data)) {
```
Comment (Subv) [inline src/core/file_sys/archive_romfs.cpp]: romfs_data.get()?

File: src/core/file_sys/archive_sdmc.cpp [hunk 3]
```diff
@@ -16,17 +17,18 @@
...
 namespace FileSys {
 
-Archive_SDMC::Archive_SDMC(const std::string& sdmc_directory) : DiskArchive(sdmc_directory) {
+ArchiveFactory_SDMC::ArchiveFactory_SDMC(const std::string& sdmc_directory) : sdmc_directory(sdmc_directory) {
     LOG_INFO(Service_FS, "Directory %s set as SDMC.", sdmc_directory.c_str());
 }
 
-bool Archive_SDMC::Initialize() {
+bool ArchiveFactory_SDMC::Initialize() {
+    // TODO(yuriks): Shouldn't these things run in Format?
```
Comment (Subv) [inline src/core/file_sys/archive_sdmc.cpp]: I don't think games are allowed to format the SD card

**Comment by Subv** [comment]
LGTM after my comments, I'll make a PR to fix that disabled code

**Comment by Subv** [comment]
#544

File: src/core/file_sys/archive_romfs.cpp (hunk 2 above)
Comment (yuriks) [inline src/core/file_sys/archive_romfs.cpp]: Hm? That gives me a pointer.

**Comment by yuriks** [comment]
I rolled up @Subv's fixes into this PR.

**Comment by Subv** [comment]
LGTM

**Comment by purpasmart96** [comment]
LGTM

File: src/core/file_sys/archive_backend.h [hunk 4]
```diff
@@ -260,4 +246,29 @@ class ArchiveBackend : NonCopyable {
     virtual std::unique_ptr<DirectoryBackend> OpenDirectory(const Path& path) const = 0;
 };
 
+class ArchiveFactory : NonCopyable {
+public:
+    virtual ~ArchiveFactory() {
+    }
+
+    /**
+     * Get a descriptive name for the archive (e.g. "RomFS", "SaveData", etc.)
```
Comment (bunnei) [inline src/core/file_sys/archive_backend.h]: Missing an @return

Comment (yuriks) [inline src/core/file_sys/archive_backend.h]: Isn't it blindingly obvious?

File: src/core/file_sys/archive_backend.h [hunk 5]
```diff
@@ -260,4 +246,29 @@ class ArchiveBackend : NonCopyable {
...
+     */
+    virtual std::string GetName() const = 0;
+
+    /**
+     * Tries to open the archive of this type with the specified path
+     * @param path Path to the archive
+     * @return ResultCode of the operation
```
Comment (yuriks) [inline src/core/file_sys/archive_backend.h]: Wrong description for the return.

File: src/core/hle/service/fs/archive.cpp [hunk 6]
```diff
@@ -187,27 +164,15 @@ class File : public Kernel::Session {
...
-public:
-    Directory(std::unique_ptr<FileSys::DirectoryBackend>&& backend, const FileSys::Path& path)
-            : path(path), backend(std::move(backend)) {
-    }
-
-    std::string GetName() const override { return "Directory: " + path.DebugStr(); }
-
-    FileSys::Path path; ///< Path of the directory
-    std::unique_ptr<FileSys::DirectoryBackend> backend; ///< File backend interface
+    cmd_buff[1] = 0; // No error
```
Comment (bunnei) [inline src/core/hle/service/fs/archive.cpp]: RESULT_SUCCESS.raw()

File: src/core/hle/service/fs/archive.cpp [hunk 7]
```diff
@@ -236,29 +201,31 @@ class Directory : public Kernel::Session {
             ResultCode error = UnimplementedFunction(ErrorModule::FS);
             cmd_buff[1] = error.raw; // TODO(Link Mauve): use the correct error code for that.
             return MakeResult<bool>(false);
-        }
-        cmd_buff[1] = 0; // No error
-        return MakeResult<bool>(false);
     }
-};
+    cmd_buff[1] = 0; // No error
```
Comment (bunnei) [inline src/core/hle/service/fs/archive.cpp]: RESULT_SUCCESS.raw()

**Comment by bunnei** [comment]
LGTM

**Comment by bunnei** [comment]
(other than the three very minor comments)

**Comment by yuriks** [comment]
Updated.

File: src/core/file_sys/archive_backend.h (hunk 4 above)
Comment (Subv) [inline src/core/file_sys/archive_backend.h]: Can't you add the @return at the start of this sentence without duplicating it?

Comment (yuriks) [inline src/core/file_sys/archive_backend.h]: I'm pretty sure Doxygen doesn't like that. (And it would result in terrible formatting regardless.)

**Comment by yuriks** [comment]
Rebased

File: src/core/hle/service/fs/archive.cpp [hunk 8]
```diff
@@ -118,12 +95,12 @@ class File : public Kernel::Session {
         // Write to file...
         case FileCommand::Write:
         {
-            u64 offset  = cmd_buff[1] | ((u64) cmd_buff[2]) << 32;
-            u32 length  = cmd_buff[3];
-            u32 flush   = cmd_buff[4];
+            u64 offset = cmd_buff[1] | ((u64)cmd_buff[2]) << 32;
+            u32 length = cmd_buff[3];
+            u32 flush = cmd_buff[4];
```
Comment (purpasmart96) [inline src/core/hle/service/fs/archive.cpp]: Why remove the alignment?

Comment (yuriks) [inline src/core/hle/service/fs/archive.cpp]: It was incorrect, so I guess @Subv just removed it outright. I'm not a huge fan of alignment everywhere anyway.

**Comment by bunnei** [comment]
OSX on Travis has been broken all day... Going to merge anyway.

File: src/core/hle/service/fs/archive.cpp (hunk 8 above)
Comment (purpasmart96) [inline src/core/hle/service/fs/archive.cpp]: meh...

Comment (Subv) [inline src/core/hle/service/fs/archive.cpp]: This wasn't intentional, it's VS autoformatting the code when it's copied

**Pull request merged**